File: bindings/go/examples/move_functions.go
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package main

import (
	sdk "bindings/iota_sdk_ffi"
	"fmt"
	"log"
)

func isNilError(err error) bool {
	if sdkErr, ok := err.(*sdk.SdkFfiError); ok {
		return sdkErr == nil
	}
	return false
}

func main() {
	client := sdk.GraphQlClientNewDevnet()

	packageAddress, err := sdk.AddressFromHex("0x3ec4826f1d6e0d9f00680b2e9a7a41f03788ee610b3d11c24f41ab0ae71da39f")
	if err != nil {
		log.Fatalf("Failed to parse address: %v", err)
	}
	
	packageOpt, err := client.Package(packageAddress, nil)
	if !isNilError(err) {
		log.Fatalf("Failed to get package: %v", err)
	}
	if packageOpt == nil {
		log.Fatalf("Missing package: %v", err)
	}
	pkg := *packageOpt

	for moduleId, _ := range pkg.Modules() {
		moduleOpt, err := client.NormalizedMoveModule(
			packageAddress,
			moduleId.AsStr(),
			nil,
			nil,
			nil,
			nil,
			nil,
		)
		if !isNilError(err) {
			log.Fatalf("Failed to get module: %v", err)
		}
		if moduleOpt == nil {
			log.Fatalf("Module: %v not found", moduleId.AsStr())
		}
		module := *moduleOpt
		if module.Functions != nil {
			fmt.Println("Module:", moduleId.AsStr())
			for _, fun := range module.Functions.Nodes {
				fmt.Println("- ", fun.String())
			}
			fmt.Println()
		}
	}
}

File: bindings/typescript/examples/move_functions.ts
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { GraphQlClient, Address } from "../lib";

async function main() {
  const client = GraphQlClient.newTestnet();

  const packageAddress = Address.fromHex(
    "0x6f727ea576a00036657fff0ae3a6d7c8171b178bf35112d6b83b2a6272cc5f0d",
  );

  const pkg = await client.package_(packageAddress, undefined);
  if (pkg === undefined) {
    throw new Error("missing package");
  }

  for (const [moduleId] of pkg.modules()) {
    const mod = await client.normalizedMoveModule(packageAddress, moduleId.asStr());
    if (mod === undefined) {
      console.log(`module \`${moduleId.asStr()}\` not found`);
      return;
    }
    if (mod.functions !== undefined) {
      console.log(`Module: ${moduleId.asStr()}`);
      for (const fun of mod.functions.nodes) {
        console.log(`- ${String(fun)}`);
      }
      console.log();
    }
  }
}

main();
